README.md
#Amabilis v1.0

Amabilis is a stand-alone executable program for use by forestry professionals
and small forest landowners alike. Amabilis is able to virtually cruise twenty-two
different species of trees on the west coast of the US. These species hold commercial 
significance to the timber industry.


Amabilis reads minimal plot data to generate current stand conditions and virtually cruises
the trees for log grades and volume. Amabilis uses minimum top DIB (Diameter inside Bark) and 
minimum log length standards for optimal log grading according to the 
"Official Rules for the Log Scaling and Grading Bureaus". It should be noted that, as Amabilis 
cannot account for log defect or knot density, these grades are the "optimal grades' and may
not reflect the true conditions of the logs.


Amabilis requires six data and one optional datum. The six required data are...

    1) Stand Name-	Can be any meaningful string of characters for example: 'EX1'.

    2) Plot Number-	The plot number within the stands, needs to be an integer.

    3) Tree Number-	The tree number within the plot, needs to be an integer.

    4) Species- 	A two character string representing the Species Codes available, for example
        'DF' for Douglas-fir. Available species and codes can be found in the Species
        drop-down menu within the 'Data Sheet Tips' area of Amabilis.

    5) DBH-		Diameter at Breast Height. This needs to be any number greater than zero. Note Amabilis 
        does not cruise trees with less than 6.0 inches DBH, but will keep them in the count 
        for stand inventory.

    6) Plot Factor- For Variable Radius plots enter the Basal Area Factor (BAF).
        For Fixed Radius plots enter the negative-inverse of the plot area, for example a
        1/30th acre plot is -30.

    The optional datum is Total Height of the tree, however this is only optional after one Total Height is 
    recorded for the stand. Total Height data are used to get an average Height to Diameter Ratio (HDR) for 
    the stand, which fills in calculated heights for trees with missing height data.



The data input side of Amabilis starts in one of two ways...

	1)  	The user can manually enter data into the built-in formatted spreadsheet.
		This can be done by navigating to the 'File' menu and selecting 'New Sheet'.

	2) 	The user can load in a pre-filled CSV file. 
		This can be done by navigating to the 'File' menu and selecting 'Browse for CSV...'
			*This CSV file should be formatted correctly*
			-To generate a blank template CSV, navigate to the 'CSV' menu and select
			'Save New CSV...'


After completion of the data entry the user has a couple options...

	1)	The user can save the data to a new CSV file.
		This can be done by navigating to the 'CSV' menu and selecting 'Save New CSV...'.

	2)	The user can append to the current CSV file.
		This can be done by navigating to the 'CSV menu and selecting 'Append Current CSV'.

	3)	The user can append to another CSV file not currently being worked on in Amabilis.
		This can be done by navigating to the 'CSV' menu and selecting 'Append Other CSV...'.


Finally the user can use the data processing tools of Amabilis. To do this navigate to the menu
'Process Data' and select 'Go To Processing'. This will open a new window containing options 
for the tools available.

Before saving or appending CSVs, or opening the Data Processing window. Amabilis will check the data for errors.
These errors could be 'at least one height needed per stand', 'missing required data' or 'value errors' 
according to data type. If there are errors, Amabilis will return you to the data entry window and highlight the 
cells with errors.


The Data Processing window contains three main tools View Report, Print Report, and exporting to FVS databases.

	  1)	View Report and Print Report
		
      -To view or print the report, click the button with the stand's name you'd like to examine, OR click 
      the button 'All Stands' to generate a report for all stands within the CSV.

      -Enter the Preferred Log Length for the stand, in forestry this number is typically 32 or 40 feet.

      -Enter the Minimum Log Length for the stand, in forestry this number is typcially 8 feet or 16 feet, but 
      this can be set to zero for maximum scaling.

      -Click the 'View Report' button to view the report in-app or click the 'Print Report' button to export the 
      report to a PDF file.

	  2) 	Exporting to FVS Databases.
		
		  FVS is a program created by the US Forest Service and hosts a suite of tools for manipulating and projecting
      forest data. There are two main versions of FVS - FVS Legacy and FVS Modern. FVS Legacy requires a 
      locations file 'Suppose.loc' and a Microsoft Access database. FVS Modern can take databases in three formats, 
      Access, SQLite3 (.db) and Excel. Amabilis will export FVS formatted databases in all three types.

      To get started in exporting to the databases navigate to the 'Export to FVS' menu and select which 
      database to export to, or select 'Export to Multiple databases' to export multiple.

      For each stand in the user's CSV file, there are eight required data and two optional data to input.
      The required data are...
      1) Variant-		The variant locale for which FVS produces its models.
      2) Forest-		Each Variant includes Forest Codes of the forests within the Forest Service.
      3) Region-		The Forest Codes are within one of the ten regions of the Forest Service.
      4) Inventory Year-	The year that the stand inventory was taken.	
      5) Age-			The approximate age of the stand in years.
      6) Elevation-		The approximate or average elevation of the stand in feet above sea level.
      7) Site Species-	The Species Code for the species used in the site index calculation.
      8) Site Index-		The site index for the majority of the stand.

      The optional data are...
      1) Latitude		Example - 45.6018
      2) Longitude		Example - -123.4371
      These can be left blank, to which Amabilis has built-in lat/longs for the geographic centers of each region.

      When the button in the bottom left changes from 'Next Stand' to 'Export to Database' all stands have been completed
      and are ready to be exported to databases for use in FVS.





Further versions of Amabilis...
Likely version 2.0 will be a major update to include the ability to process full cruise data in addition to
the quick inventory data available currently.


Amabilis is produced by Pipeline Forestry of Washington State.
To contact Pipeline.

    Email - zachbeebe@pipelineforestry.com
    Phone - 425-931-8214
    Website - pipelineforestry.com


For tree data calculations, Amabilis uses a PYPI package also created by Pipeline Forestry, called timberscale.
For more information on timberscale goto...

https://github.com/zacharybeebe/timberscale

OR

https://pypi.org/project/timberscale/

		




CHANGELOG.txt
Change Log
==========

1.0 (02/01/2021)
------------------
-First Release


source_code/report.py
from timberscale import Timber
import csv
import math

##### REPORT DATA MODULE
class Report(object):
    LOG_RANGE_LIST = [["40+ ft", range(41, 121)], ["31-40 ft", range(31, 41)], ["21-30 ft", range(21, 31)],
                      ["11-20 ft", range(11, 21)], ["1-10 ft", range(1, 11)]]


    def __init__(self, CSV, Stand_to_Examine, Plots, Pref_Log, Min_Log):
        self.csv = CSV
        self.stand = Stand_to_Examine.upper()
        self.plots = Plots
        self.plog = Pref_Log
        self.mlog = Min_Log

        self.species_list = []
        self.summary_conditions = []
        self.summary_logs = []

        self.conditions_dict = {}
        self.logs_dict = {}

        self.report()

    def report(self):
        AVG_HDR, self.species_list = self.get_HDR_Species()

        ## MAIN READ AND TREE (TIMBER CLASS) INITIALIZATION
        with open(self.csv, 'r') as tree_data:
            
            tree_data_reader = csv.reader(tree_data)
            next(tree_data_reader)
            for line in tree_data_reader:
                if line[0] == "":
                    break
                elif str(line[0]).upper() != self.stand:
                    next
                else:
                    SPECIES = str(line[3]).upper()
                    DBH = float(line[4])
                    if line[5] == "":
                        HEIGHT = int(round(AVG_HDR * (DBH/12),0))
                    else:
                        HEIGHT = int(float(line[5]))
                    PLOT_FACTOR = float(line[6])

                    if DBH >= 6.0:
                        tree = Timber(SPECIES, DBH, HEIGHT)

                        merch_dib = tree.merch_dib()
                        if merch_dib < 5:
                            merch_dib = 5
                            
                        single = tree.tree_single(merch_dib, self.plog, self.mlog)
                        tree_per_acre = tree.tree_acre(merch_dib, self.plog, self.mlog, PLOT_FACTOR)
                        log_per_acre = tree.log_acre(merch_dib, self.plog, self.mlog, PLOT_FACTOR)

                        self.summary_conditions.append([single['SPP'][0],
                                                       [tree_per_acre['TPA'], tree_per_acre['BA_AC'], tree_per_acre['RD_AC'],
                                                        single['T_HGT'], single['HDR'], single['VBAR'], tree_per_acre['BF_AC'],
                                                        tree_per_acre['CF_AC']]])


                        self.summary_logs.append(self.get_log_list(single['SPP'][0], log_per_acre))

                    else:
                        tree = Timber(SPECIES, DBH, HEIGHT)

                        self.summary_conditions.append([tree.SPP,
                                                       [tree.get_TPA(PLOT_FACTOR), tree.get_BA_acre(PLOT_FACTOR),
                                                        tree.get_RD_acre(PLOT_FACTOR), tree.HGT, tree.HDR, 0, 0, 0]])
                        


        ## SUMMARY STATISTICS
       
        self.conditions_dict = self.get_conditions_dict()
        self.logs_dict = self.get_logs_dict()

        return


    def get_HDR_Species(self):
        HDR_LIST = []
        SPECIES_LIST = []
        
        with open(self.csv, 'r') as tree_data:
            tree_data_reader = csv.reader(tree_data)
            next(tree_data_reader)

            for line in tree_data_reader:
                if line[0] == "":
                    break
                elif str(line[0]).upper() != self.stand:
                    next
                else:
                    SPP = str(line[3]).upper()
                    if SPP not in SPECIES_LIST:
                        SPECIES_LIST.append(SPP)

                    if line[5] != "":
                        DBH = float(line[4])
                        HEIGHT = float(line[5])
                        HDR_LIST.append(HEIGHT / (DBH / 12))

        AVG_HDR = round(sum(HDR_LIST) / len(HDR_LIST), 2)
        return AVG_HDR, SPECIES_LIST



    def get_log_list(self, Species, Log_Dict):
        master = [Species]
        for key in Log_Dict:
            rng = ""
            for ranges in self.LOG_RANGE_LIST:
                if Log_Dict[key]['L_LGT'] in ranges[1]:
                    rng = ranges[0]

            temp_list = [Log_Dict[key]['L_GRD'][0], rng, Log_Dict[key]['L_CT_AC'],
                         Log_Dict[key]['L_BF_AC'], Log_Dict[key]['L_CF_AC']]

            master.append(temp_list)

        return master



    def get_conditions_dict(self):
        # ORDER OF INITAL SPP LIST - [0SPPCOUNT, 1TPA, 2BA_AC, 3RD_AC, 4T_HGT, 5HDR, 6VBAR, 7BF_AC, 8CF_AC]
        #                             After Pop SPPCOUNT and Add QMD to 2 index
        # ORDER OF FINAL SPP LIST - [0TPA, 1BA_AC, 2QMD, 3RD_AC, 4T_HGT, 5HDR, 6VBAR, 7BF_AC, 8CF_AC]
        master = {}
        totals_temp = [0, 0, 0, 0, 0, 0, 0, 0, 0]
        
        for spp in self.species_list:
            master[spp] = [0, 0, 0, 0, 0, 0, 0, 0, 0]

        for data in self.summary_conditions:
            spp = data[0]
            master[spp][0] += 1
            totals_temp[0] += 1
            for i in range(1, len(data[1]) + 1):
                master[spp][i] += data[1][i - 1]
                totals_temp[i] += data[1][i - 1]

        master["TOTALS"] = totals_temp
        for key in master:
            sums = [1, 2, 3, 7, 8]
            for i in range(1, len(master[key])):
                if i in sums:
                    master[key][i] = master[key][i] / self.plots
                else:
                    master[key][i] = master[key][i] / master[key][0]
            master[key].pop(0)
            master[key].insert(2, math.sqrt((master[key][1] / master[key][0]) / .005454))

        return master


    def get_logs_dict(self):
        log_rng = ["40+ ft", "31-40 ft", "21-30 ft", "11-20 ft", "1-10 ft", 'TGRD']
        master = {}

        # Formatting Species into main keys
        for spp in self.species_list:
            master[spp] = {}

        master['TOTALS'] = {}

        # Formatting Grades and Ranges in correct order, as nested dicts of Species and Totals
        for key in master:
            for grade in Timber.GRADE_NAMES:
                master[key][grade] = {}
                for rng in log_rng:
                    master[key][grade][rng] = [0, 0, 0]
            master[key]['TTL'] = {}
            for rng in log_rng:
                master[key]['TTL'][rng] = [0, 0, 0]

        # Adding data to Master Dict
        for data in self.summary_logs:
            spp = data[0]
                
            for i in range(1, len(data)):
                grade, rng = data[i][0], data[i][1]                

                for j in range(2, len(data[i])):
                    master[spp][grade][rng][j - 2] += (data[i][j] / self.plots)
                    master[spp][grade]['TGRD'][j - 2] += (data[i][j] / self.plots)
                    
                    master[spp]['TTL'][rng][j - 2] += (data[i][j] / self.plots)
                    master[spp]['TTL']['TGRD'][j - 2] += (data[i][j] / self.plots)
                    
                    master['TOTALS'][grade][rng][j - 2] += (data[i][j] / self.plots)
                    master['TOTALS'][grade]['TGRD'][j - 2] += (data[i][j] / self.plots)

                    master['TOTALS']['TTL'][rng][j - 2] += (data[i][j] / self.plots)
                    master['TOTALS']['TTL']['TGRD'][j - 2] += (data[i][j] / self.plots)


        # Removing any Grades that have zero data
        ax_list = []
        for key in master:
            for grade in master[key]:
                count = 0
                for rng in master[key][grade]:
                    count += master[key][grade][rng][0]

                if count == 0:
                    ax_list.append((key, grade))

        for ax in ax_list:
            del master[ax[0]][ax[1]]
                    
        return master














source_code/parameters.py
##### PARAMETERS SCRIPT

## IMPORTS FOR ALL SCRIPTS
import tkinter as t
from tkinter import ttk, filedialog, messagebox
from ttkthemes import themed_tk as tk
from icon_bitmap import ICON
from win32com.client import Dispatch
import pyodbc as db
import openpyxl as xl
from openpyxl.styles import Alignment
import sqlite3 as lite
import os
from report import *
from pdf import *



PROGRAM = 'Amabilis'
VERSION = ' v1.0'

#### MAIN PROGRAM FORMATTING
BLACK = "#000000"
GREY = "#D7D7D7"
RED = "#FF0000"
DARKRED = "#A10C0C"

FORESTGREEN = "#418A52"

WIDTH = 1500
HEIGHT = 750

SCREEN_POS_RIGHT = 5
SCREEN_POS_DOWN = 5

#'Calibri'
#'Candara'

font_name = 'Calibri'

font8Cb = (font_name, "8", "bold")
font10Cb = (font_name, "10", "bold")
font11Cb = (font_name, "11", "bold")
font12Cb = (font_name, "12", "bold")
font14Cb = (font_name, "14", "bold")
font18Cb = (font_name, "18", "bold")
font36Cb = (font_name, "36", "bold")


METRIC_LIST = ["STAND", "PLOT #", "TREE #", "SPECIES", "DBH", "TOTAL HGT", "PLOT FACTOR"]


METRIC_DICT = {"STAND": ["Enter the name of the stand for example: 'EX1'."],
               "PLOT #": ["Enter the plot number within the stand."],
               "TREE #": ["Enter the tree number within the plot."],
               "SPECIES": "",
               "DBH": ["Enter the DBH of the tree."],
               "TOTAL HGT": ["Enter the total height of the tree.", "Note that only one height is needed per stand",
                             "but more heights will produce more reliable data."],
               "PLOT FACTOR": ["Variable-Radius Plots:",  "\tEnter Basal Area Factor", "",
                               "Fixed-Radius Plots:", "\tEnter the negative-inverse of", "\tthe radius (1/30th ac = -30)."]}




## ACCESS DATA TYPES
vc = 'varchar'
dbl = 'double'
sht = 'short'
lt = 'longtext'

## SQLITE3 DATA TYPES
tt = 'TEXT'
it = 'INTEGER'
rl = 'REAL'

## EXCEL SHEET NAMES
FVS_G = 'FVS_GroupAddFilesAndKeywords'
FVS_S = 'FVS_StandInit'
FVS_T = 'FVS_TreeInit'


## FVS TABLES AND COLUMN VALUES
FVS_KEYWORDS = "Database\r\nDSNIn\r\n{FILLER_DB_NAME}\r\nStandSQL\r\nSELECT *\r\nFROM FVS_StandInit\r\nWHERE Stand_ID = '%StandID%'\r\nEndSQL\r\nTreeSQL\r\nSELECT *\r\nFROM FVS_TreeInit\r\nWHERE Stand_ID = '%StandID%'\r\nEndSQL\r\nEND"


EXCEL_KEYWORDS = "Database\nDSNIn\n{FILLER_DB_NAME}\nStandSQL\nSELECT * FROM FVS_StandInit WHERE Stand_ID =\n'%StandID%'\nEndSQL\nTreeSQL\nSELECT * FROM FVS_TreeInit WHERE Stand_ID =\n'%StandID%'\nEndSQL\nEND"



FVS_GROUPS = {FVS_G: [['Groups', vc, tt, 'All_Stands'],
                      ['Addfiles', vc, tt, ''],
                      ['FVSKeywords', lt, tt, FVS_KEYWORDS]]}


FVS_STANDS = {FVS_S: [['Stand_ID', vc, tt], ['Variant', vc, tt], ['Inv_Year', sht, it],
                      ['Groups', vc, tt], ['Latitude', dbl, rl], ['Longitude', dbl, rl],
                      ['Region', sht, it], ['Forest', sht, it], ['Age', sht, it], ['ElevFt', sht, it],
                      ['Basal_Area_Factor', dbl, rl], ['Brk_DBH', dbl, rl], ['Num_Plots', sht, it],
                      ['Site_Species', vc, tt], ['Site_Index', sht, it]]}


FVS_TREES = {FVS_T: [['Stand_ID', vc, tt], ['StandPlot_ID', vc, tt], ['Plot_ID', sht, it],
                     ['Tree_ID', sht, it], ['Tree_Count', sht, it], ['History', sht, it],
                     ['Species', vc, tt], ['DBH', dbl, rl], ['Ht', dbl, rl]]}


STANDS_INPUT = [['Variant', 'Forest Code', 'Region Code'],
                ['Inventory Year', 'Latitude', 'Longitude', 'Age', 'Elevation', 'Site Species', 'Site Index']]



## FOREST CODES FOR EACH VARIANT
VARIANTS_LOCS = {'AK': ['Alaska', [1002, 1003, 1004, 1005]],
                 'BM': ['Blue Mountains', [604, 607, 614, 616, 619]],
                 'CA': ['Inland California', [505, 506, 508, 511, 514, 518, 610, 611]],
                 'CI': ['Central Idaho', [117, 402, 406, 412, 413, 414]],
                 'CR': ['Central Rockies', [202, 203, 204, 206, 207, 209, 210, 212, 213, 214, 215, 301, 302, 303, 304,
                                            305, 306, 307, 308, 309, 310, 312]],
                 'CS': ['Central States', [905, 908, 912]],
                 'EC': ['East Cascades', [603, 606, 608, 613, 617, 699]],
                 'EM': ['Eastern Montana', [102, 108, 109, 111, 112, 115]],
                 'IE': ['Inland Empire', [103, 104, 105, 106, 110, 113, 114, 116, 117, 118, 621]],
                 'LS': ['Lake States', [902, 903, 904, 906, 907, 909, 910, 913, 924]],
                 'NC': ['Klamath Mountains', [505, 510, 514, 611]],
                 'NE': ['Northeast', [914, 919, 920, 921]],
                 'PN': ['Pacific Northwest', [609, 612]],
                 'SN': ['Southern', [80101, 80103, 80104, 80105, 80106, 80107,
                                     80211, 80212, 80213, 80214, 80215, 80216, 80217,
                                     80301, 80302, 80304, 80305, 80306, 80307, 80308,
                                     80401, 80402, 80403, 80404, 80405, 80506,
                                     80501, 80502, 80504, 80505, 80506,
                                     80601, 80602, 80603, 80604, 80605,
                                     80701, 80702, 80704, 80705, 80706, 80717,
                                     80801, 80802, 80803, 80804, 80805, 80806,
                                     80811, 80812, 80813, 80814, 80815, 80816,
                                     80901, 80902, 80903, 80904, 80905, 80906,
                                     80907, 80908, 80909, 80910, 80911, 80912,
                                     81001, 81002, 81003, 81004, 81005, 81006, 81007,
                                     81102, 81103, 81105, 81106, 81107, 81108, 81109, 81110, 81111,
                                     81201, 81202, 81203, 81205,
                                     81301, 81303, 81304, 81307, 81308, 905, 908]],
                 'SO': ['SC Oregon NE California', [505, 506, 509, 511, 601, 602, 620]],
                 'TT': ['Tetons', [403, 405, 415, 416]],
                 'UT': ['Utah', [401, 407, 408, 410, 418, 419]],
                 'WC': ['Western Cascades', [603, 605, 606, 610, 615, 618]],
                 'WS': ['Western Sierra Nevada', [503, 511, 513, 515, 516, 517]]}


## DEFAULT LAT AND LONG FOR EACH VARIANT (Chosen from center of Variant geographic region)
LAT_LONG_DEFAULT = {'AK': (65.3488, -150.9184),
                    'BM': (44.9964, -118.3989),
                    'CA': (39.9440, -122.2660),
                    'CI': (43.7716, -115.1908),
                    'CR': (36.7042, -106.0942),
                    'CS': (39.6401, -90.5375),
                    'EC': (47.7250, -120.2885),
                    'EM': (47.6658, -105.1274),
                    'IE': (47.7250, -116.1576),
                    'LS': (45.4606, -90.4496),
                    'NC': (41.9355, -123.7602),
                    'NE': (41.7718, -75.0688),
                    'PN': (45.6018, -123.4371),
                    'SN': (34.5446, -87.9665),
                    'SO': (42.8761, -119.3443),
                    'TT': (42.4537, -110.5844),
                    'UT': (39.1391, -114.8926),
                    'WC': (45.9024, -121.4266),
                    'WS': (37.3916, -118.6263)}

                                                  





        
















    

